HeterogeneousCore/AlpakaInterface/src/Backend.cc

```diff
@@ -0,0 +1,23 @@
+#include "FWCore/Utilities/interface/Exception.h"
+#include "HeterogeneousCore/AlpakaInterface/interface/Backend.h"
+
+#include <algorithm>
+#include <array>
+
+namespace {
+  constexpr const std::array<std::string_view, static_cast<short>(cms::alpakatools::Backend::size)> backendNames = {
+      {"SerialSync", "CudaAsync", "ROCmAsync", "TbbAsync"}};
+}
+
+namespace cms::alpakatools {
+  Backend stringToBackend(std::string_view name) {
+    auto found = std::find(backendNames.begin(), backendNames.end(), name);
+    if (found == backendNames.end()) {
+      cms::Exception ex("EnumNotFound");
+      ex << "Invalid backend name '" << name << "'";
+      ex.addContext("Calling cms::alpakatools::stringToBackend()");
+      throw ex;
+    }
+    return static_cast<Backend>(std::distance(backendNames.begin(), found));
+  }
```

Comment:
Could this be made `constexpr` ?
With c++20 it should be easier, but maybe also with c++17 we can manage:
```c++
  constexpr Backend stringToBackend(std::string_view name) {
    using namespace std::string_literals;
    for (short i = 0; i < static_cast<short>(Backend::size); ++i) {
      if (backendNames[i] == name)
        return static_cast<Backend>(i);
    }
    throw std::out_of_range("EnumNotFound: Invalid backend name '"s + std::string(name) + "'"s);
  }
```

I do not know if a `cms::Exception` can the used in a `constexpr` context, though.

Reply:
With `constexpr` the function would have the be defined in the header (in order to have any real impact), and that would imply defining the "string" values in the header as well. I opted here to have the "string" values private to the source file to avoid proliferating the "string" constants to many shared libraries (not sure how relevant that is nowadays though).

Do I assume correctly that you'd like to see if `stringToBackend("Nonexistent")` would result already a compilation failure (which sounds neat)? Based on quick testing that would require the calls to be along
```cpp
// constexpr specifier is necessary for compile-time evaluation
constexpr auto bk = stringToBackend("CudaAsync");
```
, in which case why not just use the enumeration (`cms::alpakatools::Backend::CudaAsync`) directly? Or do you think the string value could come from some compile-time function call(s)? I had imagined the string value would primarily originate from configuration, and therefore would be known only at run time.

> I do not know if a `cms::Exception` can the used in a `constexpr` context, though.

As far as I can tell the standard exception classes are not `constepr`, and already the `throw` makes the compiler to emit the error message. So `throw cms::Exception("EnumNotFound") << "..";` might work in a `constexpr` function (I didn't test though), but I'd expect the pattern of constructing constructing the exception object, calling non-`constexpr` functions, and only then throwing, to not work.